autofit/graphical/messages/transform.py

```diff
@@ -0,0 +1,339 @@
+
+
+from abc import ABC, abstractmethod
+from autofit.graphical.utils import numerical_jacobian
+from autofit.mapper.operator import MultiVecOuterProduct
+from functools import wraps
+from typing import Type, Union, Tuple
+
+import numpy as np
+from scipy.special import ndtr, ndtri
+from scipy.stats._continuous_distns import _norm_pdf
+
+from ...mapper.operator import (
+    DiagonalMatrix,
+    LinearOperator,
+    ShermanMorrison
+)
+from ..factor_graphs import transform
+
+
+class AbstractDensityTransform(ABC):
+    """
+    Apply transforms and inverse transforms, 
+
+    def f(x):
+        return 2*x
+    """
+    @abstractmethod
+    def transform(self, x):
+        pass
+
+    @abstractmethod
+    def inv_transform(self, x):
+        pass
+
+    @abstractmethod
+    def jacobian(self, x: np.ndarray) -> LinearOperator:
+        pass
+
+    @abstractmethod
+    def log_det(self, x: np.ndarray) -> np.ndarray:
+        pass
+
+    @abstractmethod
+    def log_det_grad(self, x: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
+        pass
+
+    def log_det_hess(self, x: np.ndarray) -> np.ndarray:
+        return numerical_jacobian(
+            x, lambda x: self.log_det_grad(x)[1].sum(0)
+        )
+
+    def transform_det(self, x) -> Tuple[np.ndarray, np.ndarray]:
+        return self.transform(x), self.log_det(x)
+
+    def transform_jac(self, x) -> Tuple[np.ndarray, LinearOperator]:
+        return self.transform(x), self.jacobian(x)
+
+    def transform_det_jac(
+        self, x
```

Comment:
Does it definitely make sense to implement this function? Kind of feels like it just obfuscates underlying function calls.

Actually quite a few of these functions do. Could make sense if a child class is likely to provide a custom implementation I guess?

Reply:
yeah the child classes may define a custom implementation that can reuse intermediate calculation values to avoid recalculation